### PR TITLE
Use bool::then instead of then_some with function calls

### PR DESCRIPTION
**Pull request by tmpolaczyk**
I noticed that hardware benchmarks are being run even though we pass the --no-hardware-benchmarks cli flag. After some debugging, the cause is an incorrect usage of the `then_some` method.

From [std docs](https://doc.rust-lang.org/std/primitive.bool.html#method.then_some):

> Arguments passed to then_some are eagerly evaluated; if you are passing the result of a function call, it is recommended to use [then](https://doc.rust-lang.org/std/primitive.bool.html#method.then), which is lazily evaluated.

```rust
let mut a = 0;
let mut function_with_side_effects = || { a += 1; };

true.then_some(function_with_side_effects());
false.then_some(function_with_side_effects());

// `a` is incremented twice because the value passed to `then_some` is
// evaluated eagerly.
assert_eq!(a, 2);
```

This PR fixes all the similar usages of the `then_some` method across the codebase.

polkadot address: 138eUqXvUYT3o4GdbnWQfGRzM8yDWh5Q2eFrFULL7RAXzdWD

### CODE REVIEW

File: substrate/client/network/sync/src/strategy/state.rs [hunk 1]
```diff
@@ -173,7 +173,7 @@ impl<B: BlockT> StateStrategy<B> {
 		peer_id: PeerId,
 		announce: &BlockAnnounce<B::Header>,
 	) -> Option<(B::Hash, NumberFor<B>)> {
-		is_best.then_some({
+		is_best.then(|| {
```
Comment (tmpolaczyk) [inline substrate/client/network/sync/src/strategy/state.rs:176]: This one is pretty bad, it makes `on_validated_block_announce` behave as if `is_best` was always `true`

**Review by ggwpez — APPROVED** [review]
Huge facepalm moment, thanks for fixing!

**Review by shawntabrizi — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by ggwpez** [comment]
@paritytech/smart-contracts PTAL - CI needs your approval

**Review by pgherveou — APPROVED** [review]

**Review by seadanda — APPROVED** [review]

**Pull request merged**